esp-config/src/bin/esp-config/main.rs

```diff
@@ -0,0 +1,346 @@
+use std::{
+    collections::HashMap,
+    error::Error,
+    path::{Path, PathBuf},
+};
+
+use clap::Parser;
+use env_logger::{Builder, Env};
+use esp_config::{ConfigOption, Value};
+use serde::Deserialize;
+use toml_edit::{DocumentMut, Formatted, Item, Table};
+
+mod tui;
+
+const DEFAULT_CONFIG_PATH: &str = ".cargo/config.toml";
+
+#[derive(Parser, Debug)]
+#[command(author, version, about, long_about = None)]
+struct Args {
+    /// Root of the project
+    #[arg(short = 'P', long)]
+    path: Option<PathBuf>,
+
+    /// Chip
+    #[arg(short = 'C', long)]
+    chip: Option<esp_metadata::Chip>,
+
+    /// Config file - using `.cargo/config.toml` by default
+    #[arg(short = 'c', long)]
+    config_file: Option<String>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct CrateConfig {
+    name: String,
+    options: Vec<ConfigItem>,
+    checks: Option<Vec<String>>,
+}
+
+#[derive(Deserialize, Debug, Clone, PartialEq, Eq)]
+struct ConfigItem {
+    option: ConfigOption,
+    actual_value: Value,
+}
+
+fn main() -> Result<(), Box<dyn Error>> {
+    Builder::from_env(Env::default().default_filter_or(log::LevelFilter::Info.as_str()))
+        .format_target(false)
+        .init();
+
+    let args = Args::parse();
+
+    let work_dir = args.path.clone().unwrap_or(".".into());
+
+    let config_file_path =
+        work_dir.join(args.config_file.as_deref().unwrap_or(DEFAULT_CONFIG_PATH));
+    if !config_file_path.exists() {
+        return Err(format!(
+            "Config file {} does not exist or is not readable.",
+            config_file_path.display()
+        )
+        .into());
+    }
+
+    let mut configs = parse_configs(&work_dir, args.chip, args.config_file.clone())?;
+    let initial_configs = configs.clone();
+    let mut previous_config = initial_configs.clone();
+    let mut errors_to_show = None;
+
+    loop {
+        let repository = tui::Repository::new(configs.clone());
+
+        // TUI stuff ahead
+        let terminal = tui::init_terminal()?;
+
+        // create app and run it
+        let updated_cfg = tui::App::new(errors_to_show, repository).run(terminal)?;
+
+        tui::restore_terminal()?;
+
+        // done with the TUI
+        if let Some(updated_cfg) = updated_cfg {
+            configs = updated_cfg.clone();
+            apply_config(
+                &work_dir,
+                updated_cfg.clone(),
+                previous_config.clone(),
+                args.config_file.clone(),
+            )?;
+            previous_config = updated_cfg;
+        } else {
+            println!("Reverted configuration...");
+            apply_config(&work_dir, initial_configs, vec![], args.config_file.clone())?;
+            break;
+        }
+
+        if let Some(errors) = check_after_changes(&work_dir, args.chip, args.config_file.clone())? {
+            errors_to_show = Some(errors);
```

Comment:
I still think this is bad UX and we shouldn't let users select conflicting choices in the first place.

Reply:
If that is hard, maybe we could check configs after each time something is modified, and show the error right away.

Reply:
Yes I thought the same

Reply:
The comment can be resolved by removing the loop or by adding a comment why it's still needed ;)

Reply:
I kept the loop "just in case" - removed